File: JavaScript/Binary Tree - DFS/1448. Count Good Nodes in Binary Tree.js
/* 1448. Count Good Nodes in Binary Tree
Difficulty: Medium
Submiission Link: https://leetcode.com/problems/count-good-nodes-in-binary-tree/submissions/1388924615/ */

var goodNodes = function (root) {
    if (!root) return 0;

    let numGoodNodes = 0;
    const stack = [];
    stack.push([root, -Infinity]);
    while (stack.length) {
        let [node, max] = stack.pop();
        const { left, right, val } = node;
        if (max <= val) {
            numGoodNodes += 1;
            max = val;
        }

        if (left) stack.push([left, max]);
        if (right) stack.push([right, max]);
    }

    return numGoodNodes;
};

File: C#/Binary Tree - DFS/1448. Count Good Nodes in Binary Tree.cs
/* 1448. Count Good Nodes in Binary Tree
Difficulty: Medium
Submiission Link: https://leetcode.com/problems/count-good-nodes-in-binary-tree/submissions/1388925664/ */

public class Solution {
    public int goodNodes = 0;
    public int GoodNodes(TreeNode root) {
        if(root==null){
            return 0;
        }
        if(root.left ==null && root.right == null){
            return 1;
        }
        GetGoodNodeCount(root,root.val);
        return goodNodes;
    }

    public void GetGoodNodeCount(TreeNode root,int currentMax){
        if(root==null){
            return;
        }
        if(root.val>=currentMax){
            currentMax = root.val;
            goodNodes++;
        }

        GetGoodNodeCount(root.left,currentMax);
        GetGoodNodeCount(root.right,currentMax);
    }
}